internal/gadget-runner/gadgetrunner.go

```diff
@@ -0,0 +1,125 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Package gadgetrunner handles initializing gadgets and installed operators before
+handing them over to a specified runtime.
+*/
+package gadgetrunner
+
+import (
+	"context"
+	"fmt"
+
+	columnhelpers "github.com/inspektor-gadget/inspektor-gadget/internal/column-helpers"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/runtime"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+// GadgetRunner handles running gadgets by the gadget interface; it orchestrates the whole lifecycle of the gadget
+// instance and communicates with gadget and runtime.
+type GadgetRunner struct {
+	ctx          context.Context
+	id           string
+	gadget       gadgets.Gadget
+	gadgetParams params.Params
+	runtime      runtime.Runtime
+	columns      columnhelpers.Columns
+	operators    operators.Operators
+	logger       logger.Logger
+	result       []byte
+	resultError  error
+}
+
+func NewGadgetRunner(
+	ctx context.Context,
+	id string,
+	runtime runtime.Runtime,
+	gadget gadgets.Gadget,
+	columns columnhelpers.Columns,
+	logger logger.Logger,
+) *GadgetRunner {
+	return &GadgetRunner{
+		ctx:     ctx,
+		id:      id,
+		gadget:  gadget,
+		runtime: runtime,
+		columns: columns,
+		logger:  logger,
+	}
+}
+
+func (r *GadgetRunner) ID() string {
+	return r.id
+}
+
+func (r *GadgetRunner) Context() context.Context {
+	return r.ctx
+}
+
+func (r *GadgetRunner) Columns() columnhelpers.Columns {
+	return r.columns
+}
+
+func (r *GadgetRunner) Runtime() runtime.Runtime {
+	return r.runtime
+}
+
+func (r *GadgetRunner) Gadget() gadgets.Gadget {
+	return r.gadget
+}
+
+func (r *GadgetRunner) Operators() operators.Operators {
+	return r.operators
+}
+
+func (r *GadgetRunner) Logger() logger.Logger {
+	return r.logger
+}
+
+func (r *GadgetRunner) SetResult(result []byte, resultError error) {
+	r.result = result
+	r.resultError = resultError
+}
+
+func (r *GadgetRunner) GetResult() ([]byte, error) {
+	return r.result, r.resultError
+}
+
+func (r *GadgetRunner) GadgetParams() *params.Params {
+	return &r.gadgetParams
+}
+
+// RunGadget is the main function of GadgetRunner and controls the lifecycle of the gadget
+func (r *GadgetRunner) RunGadget(
+	runtimeParams params.Params,
```

Comment:
The runner already knows the runtime, why to pass this here?

Reply:
Getting the `Params()` from runtime, operator and gadget instantiates them, this passes the eventually filled-out (in our case right now by cobra) params to the runner - same applies to the gadget params.

Reply:
How about to make `Params()` always return the reference to the **same** parameters instance, then by calling `runtimeinstance.Params()` we'll get those parameters filled-out and we won't need to pass extra instances of them around?

Reply:
Hmm, in case of gadgettracermgr you might want to run several things at the same time but with different configurations (right now it would be based on what comes in via CR). While that might not apply to the runtimeParams, it will for the perGadgetParams of the operators and the gadgetParams - and then I'd prefer to use them all in the same way and not make an exception for the runtime params.

Reply:
I see. Is there a possibility to assign the filled-out parameters to the gadget instance? It's still confusing that `Params()` returns the empty parameters and then we have to carry around the filled-out parameters to different places. I feel like this brokes the encapsulation and makes the code more difficult to follow.

Reply:
We could do that for the perGadgetParams for operators and gadgetParams, but not for the rest as they might have been filled out/run earlier and operate in a bigger scope than the gadget (but might still be needed if runtime/operators haven't been initialized before). Example: the "LocalManager" will only be initialized once in gadgettracermgr (and for example the interactive cli thingy and a web backend) - using the global params. But I agree, this could be improved, I'll think about how to make it easier as well.

Reply:
Perhaps splitting the `Params` type in two can help? One that holds the param description (name, type, ..), and another one that holds the param value? Then different `Params()` methods around would return the descriptor, to make it clear that the value is passed in another place.

Reply:
Could you please give an example of how that would look like? I'm not sure I fully understand.

Reply:
Sure.

I think we could define these two types:

```golang
// ParamDesc describes a parameter. 
type ParamDesc struct {
	Key string 
	Alias string 
	///
}

// ParamValue contains a parameter with its value
type ParamValue struct {
	// ParamDesc ???

	value string 
}
```

Different `Params()` functions will be transformed to:

```golang
	ParamsDesc() params.ParamsDesc
```

It'll make it totally clear that `ParamsDesc()` returns only the description of the parameters that a given gadget / operator / runtime needs, but the values will be handled in a different path. 

My concern with the current API is:

```golang
params := gadget.Params()

// all logic that binds params to cobra / whatever 
bindToCobraOrWhateverThingWeHave(params)

// other logic

// This works
params["foo"].String()

// This one is empty.
gadget.Params()["foo"].String() 
```

TBH it's not a very big concern on my side. But I was a bit confused because parameters where passed all around and the `Params()` method on the different objects wasn't used too much. 

What do you think?

Reply:
Hmm, and then you'd just have a function on `ParamDesc` to convert it to a `ParamValue` - and then similarly for `ParamsDesc` to  `ParamsValue` (which calls the mentioned function on each ParamDesc) so that you would call `gadget.Params().AsValues()` or something like that? And that one would be used for calling stuff, then?

Reply:
Yes. I think we can embed `ParamDesc` in `ParamValue` and then we only need a `ParamDescToValue` (or similar) method that creates a `ParamValue` from a `ParamDesc`. Then, we'll use `ParamValue` to hold filled-out parameters around.

Reply:
I've implemented it this way now - but instead of `ParamValue` i left that one as `Param`, because it's the thing that is actually the parameter that you pass. Wdyt?

Also: the gadget/operator/runtime interface functions for returning the `ParamDescs` are still called Params() - should we change that as well or leave it?

Reply:
> Also: the gadget/operator/runtime interface functions for returning the ParamDescs are still called Params() - should we change that as well or leave it?

I think it's better to call them `ParamDescs` to make it totally clear what they are about given that we have `Param` and `ParamDesc` around.

Reply:
> My concern with the current API is:
> [...]
>

If I'm not missing something, we still have the same problem:

```golang
params := gadget.Params().ToParams() // BTW, it looks strange to call ToParms on a Param object.

// all logic that binds params to cobra / whatever 
bindToCobraOrWhateverThingWeHave(params)

// other logic

// This works
params["foo"].String()

// This one is empty.
gadget.Params().ToParams()["foo"].String() 
```

To fully solve it, we need to hide `ParamDesc` from the final user: `registry` in this case. To do it, we need to guarantee that the different objects (`runtime`, `operators` and `gadgets`) return only `Params` and never `ParamDesc`.

I have a proposal. What if the different objects allow the creation of a new instance of their parameters and retrieve them using an identifier like the following?

```golang
        // Returns a new instance for a run of a gadget identified with id.
	NewParams(id string) *params.Params

        // Returns a pointer to the same instance so we don't need to pass the params instance everywhere.
	GetParams(id string) *params.Params
```

The ID would be the `runner.ID` as it is known everywhere, and we could use the same ID we use for the `traceID`, as such ID identifies a run of a gadget.

This proposal also allows for avoiding passing parameter instances through multiple functions to get back to their owner. For example, we ask the operators to instantiate the gadget parameters. We move this instance from the registry to runner, then to runtime to finally give it back to the operator in `PreGadgetRun`: operator -> registry -> runner -> runtime -> operator.

I implemented it, and it doesn't look that bad. I opened a [PR](https://github.com/inspektor-gadget/inspektor-gadget/pull/1301) to quickly add comments and discuss it.

Reply:
> we still have the same problem:

```golang
// This one is empty.
gadget.Params().ToParams()["foo"].String() 
```
I don't really see a problem with that anymore tbh, as `ToParams()` implies that it does something to the Params(). Like discussed elsewhere, we should probably rename gadget.Params() to gadget.ParamDescs() to make it more clear that the result can't store values.

```golang
params := gadget.ParamDescs().ToParams()

...

// This should make it even more clear that something happens to ParamDescs()
gadget.ParamDescs().ToParams()["foo"].String() 
```

Regarding your PR:

Renaming Params() to GlobalParams() makes much more sense. Maybe we should use GlobalParams() for the runtime as well. So it's clear they only exist once and are probably read on start of the binary.

> To do it, we need to guarantee that the different objects (runtime, operators and gadgets) return only Params and never ParamDesc.

I think it's actually good to only return ParamDescs, as in some cases, you don't need the additional value (for example when giving the metadata to a documentation generator, passing it to a web-frontend or otherwise exposing it to third-parties). IMHO, there should be no need to instantiate anything (and thereby having to deal with the logic of instantiating) to just get that information - apart from the `&Gadget{}` added to the registry. This is why I much more like having ParamDesc and Param split in two now.
I think, this also overweighs removing the possible confusion that might be a result of the code mentioned above.

> This proposal also allows for avoiding passing parameter instances through multiple functions to get back to their owner.

I'd have liked to store the params instances in the respective objects themselves like you did, but anytime I tried, it somehow felt weird or led to handling that was specific to runtimes, operators or gadgets - or not generic enough.

> What if the different objects allow the creation of a new instance of their parameters and retrieve them using an identifier like the following?

Having to look up params by an ID in the gadget code adds unnecessary complexity and confusion IMHO. Before, the API was totally clear where you could get your params from (runner.GadgetParams()), now you have to know about a relationship with the trace/runner ID - and you might even access stuff that isn't bound to the current gadget run (using another ID). (btw: who cleans those maps up and when?)

Also, the `Params()` (or later `ParamDescs()`) on gadgets were the single point you could be sure about contains/returns the parameters - same for Runtime / Operators. In your PR that "feature" gets lost a bit by defining them in constructors or elsewhere.

I'd rather have or even add more complexity (and maybe even confusion) to our side of the code (the runtime, the runners, basically the glue) than on the gadgets' side - to make it as easy as possible to develop new gadgets.

Reply:
> Like discussed elsewhere, we should probably rename gadget.Params() to gadget.ParamDescs() to make it more clear that the result can't store values.
>
That would help.

> Renaming Params() to GlobalParams() makes much more sense. Maybe we should use GlobalParams() for the runtime as well. So it's clear they only exist once and are probably read on start of the binary.
>
Agree. I think it should be done on runtime as well.

> I think it's actually good to only return ParamDescs, as in some cases, you don't need the additional value (for example when giving the metadata to a documentation generator, passing it to a web-frontend or otherwise exposing it to third-parties).
>
That's a good point I didn't considered.

Reply:
I propose to resolve this discussion now that we have `Param` and `ParamDesc`. The latest push also renamed `Operator.ParamDesc` to `Operator.GlobalParamDescs` to make it more clear respect to `Operator.PerGadgetParams`.